dot/sync/chain_sync.go

```diff
@@ -53,693 +48,630 @@ func (s chainSyncState) String() string {
+				<-idleTimer.C
+			}
 
-	logger.Debugf("dispatching sync worker id %d, "+
-		"start number %d, target number %d, "+
-		"start hash %s, target hash %s, "+
-		"request data %d, direction %s",
-		w.id,
-		*w.startNumber, *w.targetNumber,
-		w.startHash, w.targetHash,
-		w.requestData, w.direction)
+			logger.Debugf("task result: peer(%s), with error: %v, with response: %v",
+				taskResult.who, taskResult.err != nil, taskResult.response != nil)
+
+			if taskResult.err != nil {
+				logger.Errorf("task result: peer(%s) error: %s",
+					taskResult.who, taskResult.err)
+
+				if !errors.Is(taskResult.err, network.ErrReceivedEmptyMessage) {
+					switch {
+					case strings.Contains(taskResult.err.Error(), "protocols not supported"):
+						cs.network.ReportPeer(peerset.ReputationChange{
+							Value:  peerset.BadProtocolValue,
+							Reason: peerset.BadProtocolReason,
+						}, taskResult.who)
+						cs.workerPool.ignorePeerAsWorker(taskResult.who)
+					default:
+						cs.workerPool.punishPeer(taskResult.who)
+					}
+				}
 
-	start := time.Now()
-	defer func() {
-		end := time.Now()
-		w.duration = end.Sub(start)
-		outcome := "success"
-		if w.err != nil {
-			outcome = "failure"
-		}
-		logger.Debugf(
-			"sync worker completed in %s with %s for worker id %d",
-			w.duration, outcome, w.id)
-		cs.resultQueue <- w
-	}()
+				cs.workerPool.submitRequest(taskResult.request, workersResults)
+				continue
+			}
 
-	reqs, err := workerToRequests(w)
-	if err != nil {
-		// if we are creating valid workers, this should not happen
-		logger.Criticalf("failed to create requests from worker id %d: %s", w.id, err)
-		w.err = &workerError{
-			err: err,
-		}
-		return
-	}
+			who := taskResult.who
+			request := taskResult.request
+			response := taskResult.response
 
-	for _, req := range reqs {
-		// TODO: if we find a good peer, do sync with them, right now it re-selects a peer each time (#1399)
-		if err := cs.doSync(req, w.peersTried); err != nil {
-			// failed to sync, set worker error and put into result queue
-			w.err = err
-			return
-		}
-	}
-}
+			if request.Direction == network.Descending {
+				// reverse blocks before pre-validating and placing in ready queue
+				reverseBlockData(response.BlockData)
+			}
 
-func (cs *chainSync) doSync(req *network.BlockRequestMessage, peersTried map[peer.ID]struct{}) *workerError {
-	// determine which peers have the blocks we want to request
-	peers := cs.determineSyncPeers(req, peersTried)
+			err := cs.validateResponse(request, response, who)
+			switch {
+			case errors.Is(err, errResponseIsNotChain):
+				logger.Criticalf("response invalid: %s", err)
+				cs.workerPool.punishPeer(taskResult.who)
+				cs.workerPool.submitRequest(taskResult.request, workersResults)
+				continue
+			case errors.Is(err, errEmptyBlockData):
+				cs.workerPool.punishPeer(taskResult.who)
+				cs.workerPool.submitRequest(taskResult.request, workersResults)
+				continue
+			case errors.Is(err, errUnknownParent):
```

Comment:
No, this error could happen since we are retrieving a chain of blocks concurrently, so it is usual for a part of the chain to arrive first then the beginning of the chain, so if the `validateResponse` returns `errUnknownParent` we just skip and include the chain it the `syncingChain` slice.